Add the item about unsafe code (#334)
paritytech/wiki: Substrate-Style-Guide.md

		qed"
	);
```

> - Unsafe code requires explicit proofs as panickers do. When introducing the unsafe code, one should

Reviewer: ```suggestion
- Unsafe code requires explicit proofs just as panickers do. When introducing unsafe code,
```

consider tradeoffs between efficiency on one hand and reliability, maintenance costs, and security on the other. 

> Here is a list of questions that may help to evaluate the tradeoffs:

Reviewer: ```suggestion
Here is a list of questions that may help evaluating the trade-off:
```

Reviewer: Both `tradeoff` and `trade-off` is correct. I prefer `trade-off`. Either way we should pick one and use it always: right now it's a mix.

>     - how much more performant or compact the resulting code will get using unsafe code, 

Reviewer: ```suggestion
    - how much more performant or compact the resulting code will be using unsafe code, 
```

Reviewer 2: I think that no-one should write unsafe code, to reduce the number of lines. This is insane. It should only be done for performance or because it can not be done in a different way.

Reviewer 3: When I first read this I interpreted "compact" from performance point of view, some tight loop that you can shave some instructions from (regardless of lines of code).

Author: Yeah, that:

> When I first read this I interpreted "compact" from performance point of view, some tight loop that you can shave some instructions from (regardless of lines of code).

and for the cases when the binary size of the code matters (i.e. in smart-contracts).

Reviewer 2: I would argue that you can write the same loop without unsafe code. 
If this is about calling some function from an external crate that is doing the unsafe code and you move this unsafe code into the loop, okay. I still vote for removing the `compact`.

Author: 1. Yes, but the point of @andresilva I believe is not that the same semantics is not expressible without `unsafe`, but that you can make such loop more performant using `unsafe`.
2. But my point about binary size still holds, right? If so, I am not sure why we should remove "compact", however, I am not married to it either, just curious.

Reviewer 2: Yeah right, keeping `compact` should not harm anyone :)

>     - how likely that invariants could be violated, 

Reviewer: ```suggestion
    - how likely is it that invariants could be violated, 
```

Although, if the dev believes there's even a slight chance that an invariant may not hold, then they'd better stay away from `unsafe`.

Reviewer 2: Invariants should always be proven with an assert.

Reviewer 4: would a debug_assert be enough?

Author: yeah, I don't want to define here what is "a slight chance" means here. I am imagining it as people would consider different tradeoffs at the same time. I am might be ok to have slight chance of not holding an invariant in a **super not important** piece of code (e.g. some sort of scripts). In this case you *just* get a bug which might be very hard to find and diagnose.

For most part of, e.g., the node this would mean we are don't to have

> even a slight chance that an invariant may not hold, 

and to enforce that it probably means that

> Invariants should always be proven with an assert.

However, I must note, that there is a chance of missing an important invariant from consideration accidentaly (in fact I think it must be one of the most common cases, but I don't have the data to back this up) and in this case the assert obviously won't work.

Reviewer 2: > would a debug_assert be enough?

No, you should make sure the invariant always holds. A panic is better than a memory exploit.

Reviewer 2: In more general, if you can not proof that you are correct and that you maybe forgot an invariant. You should not write unsafe code!

Author: The problem is: you don't not know that you don't know about that invariant (e.g., as per discussion in Riot today)

Reviewer 2: For the discussion in Riot, see [here](https://www.ralfj.de/blog/2017/06/09/mutexguard-sync.html). This was a bug in safe rust. So, this means you can not trust anything. The `uninitialized` stuff was a bug in the Rust standard library, but you probably program with good faith that the std library is "bug-free". Otherwise you are lost and can not trust anything.

TLDR: Put the invariants you are aware of, nobody will make you responsible for bugs in the Rust std library.

>     - how likely that the issue can be discovered before production (tests and tools can help but not always) and,

Reviewer: ```suggestion
    - are issues stemming from the use of unsafe code caught by existing tests/tooling
```

…but again, unless the dev can prove their use of unsafe is ok, shouldn't they stay away? Or maybe I'm misunderstanding this bullet point?

Author: Let's keep the discussion in the original thread. I am going to merge this anyway and then we can figure out what to do next based on the discussion on that thread.

>     - what are the consequences if the problems slip into production,

Reviewer 3: ```suggestion
    - what are the consequences if the problems slip into production.
```

Author: Oh, yeah, I actually started with `slips` but I then realized that "the problems" are plural and thus you can't use third-person singular here.

Tell me if I am mistaken.

Reviewer 3: Sorry, I overlooked it and thought it was `the problem` (singular). Edited the suggestion.
